packages/gatsby-plugin-functions/src/gatsby-node.ts

```diff
@@ -0,0 +1,181 @@
+    }
+  )
+
+  try {
+    await Promise.all(
+      files.map(file => {
+        const config = {
+          entry: path.join(functionsDirectory, file),
+          output: {
+            path: path.join(process.cwd(), `.cache`, `functions`),
+            filename: file.replace(`.ts`, `.js`),
+            libraryTarget: `commonjs2`,
+          },
+          target: `node`,
+          // library: "yourLibName",
+
+          mode: `production`,
+          module: {
+            rules: [
+              {
+                test: [/.js$/, /.ts$/],
+                exclude: /node_modules/,
+                loader: `babel-loader`,
+              },
+            ],
+          },
+          plugins: [new webpack.DefinePlugin(varObject)],
+          // devtool: `source-map`,
+        }
+
+        return new Promise((resolve, reject) =>
+          // if (stage === `develop`) {
+          //   webpack(config).watch({}, () => {})
+
+          //   return resolve()
+          // }
+
+          webpack(config).run((err, stats) => {
+            if (stats?.compilation?.warnings?.length > 0) {
+              reporter.warn(stats.compilation.warnings)
+            }
+
+            if (err) return reject(err)
+            const errors = stats.compilation.errors || []
+            if (errors.length > 0) return reject(stats.compilation.errors)
+            return resolve()
+          })
+        )
+      })
+    )
+  } catch (e) {
+    activity.panic(`Failed to compile Gatsby Functions.`, e)
+  }
+
+  activity.end()
+}
+
+export async function onCreateDevServer(
+  { reporter, app }: CreateDevServerArgs,
+  {
+    path: functionsDirectoryPath = DEFAULT_FUNCTIONS_DIRECTORY_PATH,
+  }: PluginOptions
+): Promise<void> {
+  const functionsGlob = `**/*.{js,ts}`
+  const functionsDirectory = path.resolve(
+    process.cwd(),
+    functionsDirectoryPath as string
+  )
+  const files = await glob(functionsGlob, { cwd: functionsDirectory })
+
+  reporter.verbose(`Attaching functions to development server`)
+
+  const knownFunctions = new Map(
+    files.map(file => [
+      urlResolve(path.parse(file).dir, path.parse(file).name),
+      file,
+    ])
+  )
+
+  app.use(
+    `/api/:functionName`,
+    multer().none(),
+    express.urlencoded({ extended: true }),
+    express.text(),
+    express.json(),
+    express.raw(),
+    (req, res, next) => {
+      const { functionName } = req.params
+
+      if (knownFunctions.has(functionName)) {
+        const activity = reporter.activityTimer(
+          `Executing function ${functionName}`
+        )
+        activity.start()
+        const compiledFunctionsDir = path.join(
+          process.cwd(),
+          `.cache`,
+          `functions`
+        )
+        const funcNameToJs = knownFunctions.get(functionName) as string
+
+        try {
+          const fn = require(path.join(compiledFunctionsDir, funcNameToJs))
+
+          const fnToExecute = (fn && fn.default) || fn
+
+          fnToExecute(req, res)
```

Comment:
`Promise.race` with a timeout controlled by an env var that can be set and defaults to 30 seconds?

Should warn if race fails